### PR TITLE
Presentation: Rework hierarchy update

### PR DESCRIPTION
**Pull request by saskliutas**
Reworked hierarchy update logic after imodel changes. Previously we were tracking UI state and tried to recreated all the necessary nodes to avoid losing expanded parts of tree in UI. The new approach is to removed from cache everything that might have changed after update and reload whole tree on the frontend. This simplifies backend code as there is no need to track UI state and generate `HierarchyUpdateRecord`s for frontend.

Part of https://github.com/iTwin/presentation/issues/57

Related TS changes: https://github.com/iTwin/itwinjs-core/pull/5219
itwinjs-core: https://github.com/iTwin/itwinjs-core/tree/presentation/rework_update_after_imodel_changes

### CODE REVIEW

File: iModelJsNodeAddon/presentation/UpdateRecordsHandler.cpp [hunk 1]
```diff
@@ -55,24 +52,10 @@ struct IModelJsECPresentationUpdateRecordsHandler::FinalReport
...
-        void push_back(HierarchyUpdateRecord record)
-            {
-            m_lastUsedTimestamp = BeTimeUtilities::GetCurrentTimeAsUnixMillis();
-            bvector<HierarchyUpdateRecord>::push_back(record);
-            }
-        uint64_t GetLastUsedTimestamp() const {return m_lastUsedTimestamp;}
-    };
-
 private:
     static uint64_t const s_partialHierarchyUpdatesLimit = 1000000; // TODO: use 1M for now, need to measure size of update records and tweak this number
```
Comment (grigasp) [inline iModelJsNodeAddon/presentation/UpdateRecordsHandler.cpp:56]: looks like `s_partialHierarchyUpdatesLimit` can be removed now

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp [hunk 2]
```diff
@@ -10,4906 +10,1413 @@ USING_NAMESPACE_BENTLEY_SQLITE_EC
...
+    auto params = AsyncHierarchyRequestParams::Create(m_db, rules->GetRuleSetId(), RulesetVariables());
+    ValidateHierarchy(params,
+        {
+        CreateInstanceNodeValidator({ a1 }),
+        });
 
-    // delete one of the instances
-    RulesEngineTestHelpers::DeleteInstance(m_db, *widget1, true);
-    m_eventsSource->NotifyECInstanceDeleted(m_db, *widget1);
+    // insertInstance
```
Comment (grigasp) [inline iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp:50]:
```suggestion
    // insert instance
```

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp [hunk 3]
```diff
@@ -10,4906 +10,1413 @@ USING_NAMESPACE_BENTLEY_SQLITE_EC
...
+    // validate hierarchy pre-update
+    auto params = AsyncHierarchyRequestParams::Create(m_db, rules->GetRuleSetId(), RulesetVariables());
+    ValidateHierarchy(params,
+        {
+        ExpectedHierarchyDef(CreateLabelGroupingNodeValidator("1", { a11, a12 }),
+            {
+            CreateInstanceNodeValidator({ a11 }),
+            CreateInstanceNodeValidator({ a12 }),
+            }),
+        CreateInstanceNodeValidator({ a21 }),
```
Comment (grigasp) [inline iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp:449]: maybe it's also worth having a test case for removing the sibling node (in this case `a21`)

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp [hunk 4]
```diff
@@ -10,4906 +10,1413 @@ USING_NAMESPACE_BENTLEY_SQLITE_EC
...
-    rule->AddSpecification(*new InstanceNodesOfSpecificClassesSpecification(1, true, false, false, false, false, false,
-        "", "RulesEngineTest:Widget", false));
-    rules->AddPresentationRule(*rule);
-
-    ChildNodeRule* childRule = new ChildNodeRule();
-    childRule->AddSpecification(*new RelatedInstanceNodesSpecification(1, false, false, false, false, false, false, false, 1, "",
-        RequiredRelationDirection_Forward, "", "RulesEngineTest:GadgetHasSprockets", "RulesEngineTest:Sprocket"));
-    rules->AddPresentationRule(*childRule);
-
-    // request f
```
Comment (grigasp) [inline iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp:1209]: let's also have a test for inserting/deleting the invisible `B` node

File: iModelJsNodeAddon/presentation/UpdateRecordsHandler.cpp (hunk 1 above)
Comment (saskliutas) [reply iModelJsNodeAddon/presentation/UpdateRecordsHandler.cpp]: Removed

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp (hunk 3 above)
Comment (saskliutas) [reply iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp:504]: Added test to check if LabelGrouping node is removed when it is only one left at the root

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp (hunk 4 above)
Comment (saskliutas) [reply iModelCore/ECPresentation/Tests/NonPublished/Integration/Hierarchies/HierarchyUpdateTests.cpp]: Added tests to check inserting/removing/ nodes in hidden hierarchy level.

**Review by grigasp — APPROVED** [review]

**Pull request merged**